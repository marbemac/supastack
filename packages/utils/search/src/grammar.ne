@preprocessor typescript

@{%

import { createLexer } from './lexer.js';

const lexer = createLexer();

%}

# https://omrelli.ug/nearley-playground/ is helpful to experiment
# Required adjustments to the content of this file if copy pasting over to playground:
#   - need to remove "@preprocessor typescript" statement
#   - need to copy createLexer function from lexer.ts, and change add "const moo = require('moo')" (es6 imports not supported)

@lexer lexer

Main -> (Qual | _ | Raw):*  {% transformMain %}

Qual ->
	QualKey QualVal:? {% ([key,val]) => ({ key, op: '=', val: val || '' }) %}
	| QualKey QualOp QualVal:? {% ([key,op,val]) => ({ key, op: op.value, val: val || '' }) %}
	| QualKey Quote QualVal:? Quote:? {% ([key,,val]) => ({ key, op: '=', val: val || '' }) %}
	| QualKey LBracket QualMultiVal:? RBracket:? {% ([key,,val]) => ({ key, op: '=', val: val || '' }) %}

QualKey -> %qualKey {% id %}

QualOp -> %qualOp {% id %}

QualVal -> %qualVal {% id %}

QualMultiVal -> (QualVal Comma:?):+ {% ([val]) => val.map((v) => v[0]) %}

Quote -> %quote {% () => null %}

LBracket -> %lbracket {% () => null %}

RBracket -> %rbracket {% () => null %}

Comma -> %comma {% () => null %}

Raw -> %string {% id %}

_ -> %space {% () => null %}

@{%

function transformMain([tokens]) {
	const noFalsey = [];
	for (const f of tokens) {
		if (f[0]) {
			noFalsey.push(f[0])
		}
	}

  return {
		tokens: noFalsey
	};
}

%}
