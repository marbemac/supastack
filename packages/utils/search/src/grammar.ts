/* eslint-disable */
// @ts-nocheck
// Generated automatically by nearley, version 2.20.1
// http://github.com/Hardmath123/nearley
// Bypasses TS6133. Allow declared but unused functions.
// @ts-ignore
function id(d: any[]): any {
  return d[0];
}
declare var qualKey: any;
declare var qualOp: any;
declare var qualVal: any;
declare var quote: any;
declare var lbracket: any;
declare var rbracket: any;
declare var comma: any;
declare var string: any;
declare var space: any;

import { createLexer } from './lexer.ts';

const lexer = createLexer();

function transformMain([tokens]) {
  const noFalsey = [];
  for (const f of tokens) {
    if (f[0]) {
      noFalsey.push(f[0]);
    }
  }

  return {
    tokens: noFalsey,
  };
}

interface NearleyToken {
  value: any;
  [key: string]: any;
}

interface NearleyLexer {
  reset: (chunk: string, info: any) => void;
  next: () => NearleyToken | undefined;
  save: () => any;
  formatError: (token: never) => string;
  has: (tokenType: string) => boolean;
}

interface NearleyRule {
  name: string;
  symbols: NearleySymbol[];
  postprocess?: (d: any[], loc?: number, reject?: {}) => any;
}

type NearleySymbol = string | { literal: any } | { test: (token: any) => boolean };

interface Grammar {
  Lexer: NearleyLexer | undefined;
  ParserRules: NearleyRule[];
  ParserStart: string;
}

const grammar: Grammar = {
  Lexer: lexer,
  ParserRules: [
    { name: 'Main$ebnf$1', symbols: [] },
    { name: 'Main$ebnf$1$subexpression$1', symbols: ['Qual'] },
    { name: 'Main$ebnf$1$subexpression$1', symbols: ['_'] },
    { name: 'Main$ebnf$1$subexpression$1', symbols: ['Raw'] },
    {
      name: 'Main$ebnf$1',
      symbols: ['Main$ebnf$1', 'Main$ebnf$1$subexpression$1'],
      postprocess: d => d[0].concat([d[1]]),
    },
    { name: 'Main', symbols: ['Main$ebnf$1'], postprocess: transformMain },
    { name: 'Qual$ebnf$1', symbols: ['QualVal'], postprocess: id },
    { name: 'Qual$ebnf$1', symbols: [], postprocess: () => null },
    {
      name: 'Qual',
      symbols: ['QualKey', 'Qual$ebnf$1'],
      postprocess: ([key, val]) => ({ key, op: '=', val: val || '' }),
    },
    { name: 'Qual$ebnf$2', symbols: ['QualVal'], postprocess: id },
    { name: 'Qual$ebnf$2', symbols: [], postprocess: () => null },
    {
      name: 'Qual',
      symbols: ['QualKey', 'QualOp', 'Qual$ebnf$2'],
      postprocess: ([key, op, val]) => ({ key, op: op.value, val: val || '' }),
    },
    { name: 'Qual$ebnf$3', symbols: ['QualVal'], postprocess: id },
    { name: 'Qual$ebnf$3', symbols: [], postprocess: () => null },
    { name: 'Qual$ebnf$4', symbols: ['Quote'], postprocess: id },
    { name: 'Qual$ebnf$4', symbols: [], postprocess: () => null },
    {
      name: 'Qual',
      symbols: ['QualKey', 'Quote', 'Qual$ebnf$3', 'Qual$ebnf$4'],
      postprocess: ([key, , val]) => ({ key, op: '=', val: val || '' }),
    },
    { name: 'Qual$ebnf$5', symbols: ['QualMultiVal'], postprocess: id },
    { name: 'Qual$ebnf$5', symbols: [], postprocess: () => null },
    { name: 'Qual$ebnf$6', symbols: ['RBracket'], postprocess: id },
    { name: 'Qual$ebnf$6', symbols: [], postprocess: () => null },
    {
      name: 'Qual',
      symbols: ['QualKey', 'LBracket', 'Qual$ebnf$5', 'Qual$ebnf$6'],
      postprocess: ([key, , val]) => ({ key, op: '=', val: val || '' }),
    },
    { name: 'QualKey', symbols: [lexer.has('qualKey') ? { type: 'qualKey' } : qualKey], postprocess: id },
    { name: 'QualOp', symbols: [lexer.has('qualOp') ? { type: 'qualOp' } : qualOp], postprocess: id },
    { name: 'QualVal', symbols: [lexer.has('qualVal') ? { type: 'qualVal' } : qualVal], postprocess: id },
    { name: 'QualMultiVal$ebnf$1$subexpression$1$ebnf$1', symbols: ['Comma'], postprocess: id },
    { name: 'QualMultiVal$ebnf$1$subexpression$1$ebnf$1', symbols: [], postprocess: () => null },
    { name: 'QualMultiVal$ebnf$1$subexpression$1', symbols: ['QualVal', 'QualMultiVal$ebnf$1$subexpression$1$ebnf$1'] },
    { name: 'QualMultiVal$ebnf$1', symbols: ['QualMultiVal$ebnf$1$subexpression$1'] },
    { name: 'QualMultiVal$ebnf$1$subexpression$2$ebnf$1', symbols: ['Comma'], postprocess: id },
    { name: 'QualMultiVal$ebnf$1$subexpression$2$ebnf$1', symbols: [], postprocess: () => null },
    { name: 'QualMultiVal$ebnf$1$subexpression$2', symbols: ['QualVal', 'QualMultiVal$ebnf$1$subexpression$2$ebnf$1'] },
    {
      name: 'QualMultiVal$ebnf$1',
      symbols: ['QualMultiVal$ebnf$1', 'QualMultiVal$ebnf$1$subexpression$2'],
      postprocess: d => d[0].concat([d[1]]),
    },
    { name: 'QualMultiVal', symbols: ['QualMultiVal$ebnf$1'], postprocess: ([val]) => val.map(v => v[0]) },
    { name: 'Quote', symbols: [lexer.has('quote') ? { type: 'quote' } : quote], postprocess: () => null },
    { name: 'LBracket', symbols: [lexer.has('lbracket') ? { type: 'lbracket' } : lbracket], postprocess: () => null },
    { name: 'RBracket', symbols: [lexer.has('rbracket') ? { type: 'rbracket' } : rbracket], postprocess: () => null },
    { name: 'Comma', symbols: [lexer.has('comma') ? { type: 'comma' } : comma], postprocess: () => null },
    { name: 'Raw', symbols: [lexer.has('string') ? { type: 'string' } : string], postprocess: id },
    { name: '_', symbols: [lexer.has('space') ? { type: 'space' } : space], postprocess: () => null },
  ],
  ParserStart: 'Main',
};

export default grammar;
