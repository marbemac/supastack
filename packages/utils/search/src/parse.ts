import nearley from 'nearley';

import grammar from './grammar.ts';
import type { LexerQualOps, LexerToken, SimpleLexerToken } from './lexer.ts';
import type { SearchString } from './types.ts';

export type LexerT = 'simple' | 'default';

/**
 * This must match the output of the nearley grammar (grammar.ne). If the grammar structure changes, update this.
 */
export interface ParseResult<T extends LexerT = LexerT> {
  tokens: ParseResultToken<T>[];
}

export interface SimpleParseResult {
  tokens: ParseResultToken<'simple'>[];
}

type ParseResultToken<T extends LexerT = LexerT> = QualifierToken<T> | RawToken<T>;

export type RawToken<T extends LexerT = LexerT> = T extends 'default'
  ? LexerToken<'string'>
  : SimpleLexerToken<'string'>;

export type QualifierToken<T extends LexerT = LexerT> = QualifierSingleToken<T> | QualifierMultiToken<T>;

export type QualifierOp = LexerQualOps | '=';

export interface QualifierSingleToken<T extends LexerT = LexerT> {
  key: T extends 'default' ? LexerToken<'qualKey'> : SimpleLexerToken<'qualKey'>;
  op: QualifierOp;
  val: T extends 'default' ? LexerToken<'qualVal'> : SimpleLexerToken<'qualVal'>;
}

export interface QualifierMultiToken<T extends LexerT = LexerT> {
  key: T extends 'default' ? LexerToken<'qualKey'> : SimpleLexerToken<'qualKey'>;
  op: LexerQualOps | '=';
  val: T extends 'default' ? LexerToken<'qualVal'>[] : SimpleLexerToken<'qualVal'>[];
}

interface ParseSearchQueryOpts {
  parser?: nearley.Parser;
}

/**
 * Some ideas in https://github.com/kach/nearley/issues/316#issuecomment-336718459 if we need to support
 * partial results in the future.
 */
export const parseSearchQuery = (input: SearchString, opts: ParseSearchQueryOpts = {}): ParseResult => {
  const parser = opts.parser || new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
  parser.feed(input);

  const output = parser.results[0];

  if (!output) {
    throw new Error('Could not parse search query.');
  }

  return output as ParseResult;
};

export const isQualifier = <T extends LexerT>(token: ParseResultToken<T>): token is QualifierToken<T> =>
  Object.prototype.hasOwnProperty.call(token, 'op');

export const isQualifierMulti = <T extends LexerT>(token: ParseResultToken<T>): token is QualifierMultiToken<T> =>
  isQualifier(token) && Array.isArray(token.val);

export const isRawToken = <T extends LexerT>(token: ParseResultToken<T>): token is RawToken<T> =>
  !token.hasOwnProperty.call(token, 'op');
