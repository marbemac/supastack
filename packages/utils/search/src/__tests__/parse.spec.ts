import nearley from 'nearley';
import { expect, it } from 'vitest';

import grammar from '../grammar.ts';
import { isRawToken, parseSearchQuery } from '../parse.ts';
import type { SearchString } from '../types.ts';
import { cases } from './cases.ts';

const prettyPrintParseResult = (input: string) => {
  const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
  const res = parseSearchQuery(input as SearchString, { parser });

  const pretty = [];
  for (const r of res.tokens) {
    if (isRawToken(r)) {
      pretty.push(`${r.value}`);
      continue;
    }

    if (Array.isArray(r.val)) {
      pretty.push(`${r.key.value} IN (${r.val.map(v => v.value).join(', ')})`);
      continue;
    }

    pretty.push(`${r.key.value} ${r.op} ${r.val.value}`);
  }

  return pretty;
};

it.each(cases)('%s', (_, input) => {
  expect({
    input,
    result: prettyPrintParseResult(input),
  }).toMatchSnapshot();
});
