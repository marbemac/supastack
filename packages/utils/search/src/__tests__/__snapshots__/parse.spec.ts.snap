// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`supports .-_ in qualifiers 1`] = `
{
  "input": "payments.plan-type_is:enterprise",
  "result": [
    "payments.plan-type_is = enterprise",
  ],
}
`;

exports[`supports @.-_ in qualifier values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": [
    "email = jane@example.com",
    "type = public-company_x",
  ],
}
`;

exports[`supports freeform search mixed throughout 1`] = `
{
  "input": "freeform1 age:5 freeform2 search",
  "result": [
    "freeform1",
    "age = 5",
    "freeform2",
    "search",
  ],
}
`;

exports[`supports multiple values on the same key 1`] = `
{
  "input": "user:[jane, john doe] release:[12.0]",
  "result": [
    "user IN (jane, john doe)",
    "release IN (12.0)",
  ],
}
`;

exports[`supports operators 1`] = `
{
  "input": "num_members:>5.0",
  "result": [
    "num_members > 5.0",
  ],
}
`;

exports[`supports qualifiers 1`] = `
{
  "input": "plan:enterprise age:5",
  "result": [
    "plan = enterprise",
    "age = 5",
  ],
}
`;

exports[`supports quoted qualifiers 1`] = `
{
  "input": "user:\\"john doe\\"",
  "result": [
    "user = john doe",
  ],
}
`;

exports[`treats freeform as a string 1`] = `
{
  "input": "freeform",
  "result": [
    "freeform",
  ],
}
`;
